majorna.1
'\" t
.\" Automatically generated by Pandoc 3.1.12.1
.\"
.TH "MAJORNA" "1" "April 2025" "" ""
.SH Majorna
\f[B]Majorna\f[R] is an \f[B]X11\f[R] and \f[B]Wayland\f[R] menu
application which takes standard input, parses it, lets the user choose
an option and sends the selected option to standard output.
.PP
Its core functionality is built around taking and parsing input and
returning output, making it a customizable and versatile utility for use
with shell or Python scripts.
.PP
It doesn\[cq]t do much by itself, but used in conjunction with projects
such as \f[CR]majorna_run\f[R], it can serve as a run launcher.
.SS Usage
On runtime, Majorna reads from standard input (stdin).
Majorna items are separated by a newline (\f[CR]\[rs]n\f[R]).
When (by default) Enter is pressed, the selected item will be piped to
stdout.
.PP
This allows things like
\f[CR]printf \[dq]Apple\[rs]nOrange\[rs]nPear\[rs]n\[dq] | majorna\f[R].
This command will spawn an Majorna window with three items, `Apple',
`Orange' and `Pear'.
The user can select an item using the keyboard or mouse and the output
can then be parsed.
This can be used in shell scripts to create interactive menus.
.PP
Note that Majorna doesn\[cq]t strictly read standard input, it can read
from files too using the \f[CR]\-lf\f[R] or \f[CR]\-\-list\-file\f[R]
argument.
.PP
On top of this, you can specify arguments to change the behavior of
Majorna.
See a list below for a list.
.SS Arguments
You may use long, descriptive arguments or the shorter arguments.
Note that none need to be specified for a basic Majorna prompt.
The most commonly used arguments are \f[CR]\-l\f[R] or
\f[CR]\-\-lines\f[R] and \f[CR]\-g\f[R] or \f[CR]\-\-columns\f[R].
All of the arguments can be defined permanently in the configuration
file.
Arguments take priority over settings defined in the config files.
.TP
\f[CR]\-mh, \-\-line\-height height\f[R]
Set Majorna line height to height
.TP
\f[CR]\-cw, \-\-center\-width width\f[R]
Set width to width when centered
.TP
\f[CR]\-l, \-\-lines lines\f[R]
Set the number of lines to lines
.TP
\f[CR]\-g, \-\-columns grid\f[R]
Set the number of grids to grid
.TP
\f[CR]\-ml, \-\-min\-lines\f[R]
Set minimum number of lines allowed to lines
.TP
\f[CR]\-gc, \-\-generate\-cache\f[R]
Generate image cache
.TP
\f[CR]\-ngc, \-\-no\-generate\-cache\f[R]
Don\[cq]t generate image cache
.TP
\f[CR]\-mc, \-\-max\-cache size\f[R]
Set max image cache size to size
.TP
\f[CR]\-cd, \-\-cache\-dir dir\f[R]
Set cache directory to dir
.TP
\f[CR]\-ix, \-\-print\-index\f[R]
Print index instead of actual text
.TP
\f[CR]\-nix, \-\-no\-print\-index\f[R]
Don\[cq]t print index instead of actual text
.TP
\f[CR]\-f, \-\-fast\f[R]
Grabs keyboard before reading stdin
.TP
\f[CR]\-r, \-\-incremental\f[R]
Print text every time a key is pressed
.TP
\f[CR]\-nr, \-\-no\-incremental\f[R]
Don\[cq]t print text every time a key is pressed
.TP
\f[CR]\-rm, \-\-require\-match\f[R]
Require that input text matches an item
.TP
\f[CR]\-nrm, \-\-no\-require\-match\f[R]
Don\[cq]t require that input text matches an item
.TP
\f[CR]\-ma, \-\-mark\-items\f[R]
Allow marking/selecting multiple items
.TP
\f[CR]\-nma, \-\-no\-mark\-items\f[R]
Don\[cq]t allow marking/selecting multiple items
.TP
\f[CR]\-F, \-\-fuzzy\f[R]
Enable fuzzy matching
.TP
\f[CR]\-NF, \-\-no\-fuzzy\f[R]
Disable fuzzy matching
.TP
\f[CR]\-R, \-\-regex\f[R]
Enable regex matching
.TP
\f[CR]\-NR, \-\-no\-regex\f[R]
Disable regex matching
.TP
\f[CR]\-P, \-\-password\f[R]
Hide characters
.TP
\f[CR]\-nP, \-\-no\-password\f[R]
Don\[cq]t hide characters
.TP
\f[CR]\-p, \-\-prompt text\f[R]
Set Majorna prompt text to text
.TP
\f[CR]\-pt, \-\-pretext text\f[R]
Set Majorna pretext to text
.TP
\f[CR]\-It, \-\-input text\f[R]
Set initial input text to text
.TP
\f[CR]\-ip, \-\-indent\f[R]
Indent items to prompt width
.TP
\f[CR]\-nip, \-\-no\-indent\f[R]
Don\[cq]t indent items to prompt width
.TP
\f[CR]\-ci, \-\-color\-items\f[R]
Color items
.TP
\f[CR]\-nci, \-\-no\-color\-items\f[R]
Don\[cq]t color items
.TP
\f[CR]\-sgr, \-\-sgr\f[R]
Interpret SGR sequences
.TP
\f[CR]\-nsgr, \-\-no\-sgr\f[R]
Display SGR sequences as text
.TP
\f[CR]\-a, \-\-alpha\f[R]
Enable alpha
.TP
\f[CR]\-na, \-\-no\-alpha\f[R]
Disable alpha
.TP
\f[CR]\-tp, \-\-allow\-typing\f[R]
Allow the user to type
.TP
\f[CR]\-nt, \-\-no\-allow\-typing\f[R]
Don\[cq]t allow typing, the user must select an option
.TP
\f[CR]\-ol, \-\-override\-lines\f[R]
Allow keybinds to override lines
.TP
\f[CR]\-oc, \-\-override\-columns\f[R]
Allow keybinds to override columns
.TP
\f[CR]\-nol, \-\-no\-override\-lines\f[R]
Don\[cq]t allow keybinds to override lines
.TP
\f[CR]\-noc, \-\-no\-override\-columns\f[R]
Don\[cq]t allow keybinds to override columns
.TP
\f[CR]\-x, \-\-x\-position x offset\f[R]
Offset Majorna x position by x offset (X11 only)
.TP
\f[CR]\-y, \-\-y\-position y offset\f[R]
Offset Majorna y position by y offset (X11 only)
.TP
\f[CR]\-n, \-\-preselect line\f[R]
Preselect line line in the list of items
.TP
\f[CR]\-z, \-\-width width\f[R]
Width of the Majorna window
.TP
\f[CR]\-nmt, \-\-normal\-mode\-text text\f[R]
Set normal mode text to text
.TP
\f[CR]\-imt, \-\-insert\-mode\-text text\f[R]
Set insert mode text to text
.TP
\f[CR]\-clon, \-\-caps\-lock\-on\-text text\f[R]
Set caps lock on text to text
.TP
\f[CR]\-clof, \-\-caps\-lock\-off\-text text\f[R]
Set caps lock off text to text
.TP
\f[CR]\-bw, \-\-border\-width width\f[R]
Set width of the border to width.
0 will disable the border (X11 only)
.TP
\f[CR]\-so, \-\-sort\f[R]
Sort matches
.TP
\f[CR]\-nso, \-\-no\-sort\f[R]
Don\[cq]t sort matches
.TP
\f[CR]\-pri, \-\-priority pri1,pri2,pri3\f[R]
Specify a list of items that take priority
.TP
\f[CR]\-s, \-\-case\-sensitive\f[R]
Use case\-sensitive matching
.TP
\f[CR]\-ns, \-\-case\-insensitive\f[R]
Use case\-insensitive matching
.TP
\f[CR]\-nm, \-\-normal\f[R]
Start Majorna in normal mode
.TP
\f[CR]\-im, \-\-insert\f[R]
Start Majorna in insert mode
.TP
\f[CR]\-t, \-\-top\f[R]
Position Majorna at the top of the screen
.TP
\f[CR]\-b, \-\-bottom\f[R]
Position Majorna at the bottom of the screen
.TP
\f[CR]\-c, \-\-center\f[R]
Position Majorna at the center of the screen
.TP
\f[CR]\-itt, \-\-item\-top\f[R]
Position items above all other elements
.TP
\f[CR]\-itb, \-\-item\-bottom\f[R]
Position items below all other elements
.TP
\f[CR]\-hm, \-\-hide\-mode\f[R]
Hide mode indicator
.TP
\f[CR]\-hit, \-\-hide\-item\f[R]
Hide items
.TP
\f[CR]\-hmc, \-\-hide\-match\-count\f[R]
Hide match count
.TP
\f[CR]\-hla, \-\-hide\-left\-arrow\f[R]
Hide left arrow
.TP
\f[CR]\-hra, \-\-hide\-right\-arrow\f[R]
Hide right arrow
.TP
\f[CR]\-hpr, \-\-hide\-prompt\f[R]
Hide prompt
.TP
\f[CR]\-hpt, \-\-hide\-pretext\f[R]
Hide pretext
.TP
\f[CR]\-hip, \-\-hide\-input\f[R]
Hide input
.TP
\f[CR]\-hpl, \-\-hide\-powerline\f[R]
Hide powerline
.TP
\f[CR]\-hc, \-\-hide\-caret, \-\-hide\-cursor\f[R]
Hide caret
.TP
\f[CR]\-hhl, \-\-hide\-highlighting\f[R]
Hide highlight
.TP
\f[CR]\-hi, \-\-hide\-image\f[R]
Hide image
.TP
\f[CR]\-hcl, \-\-hide\-caps\f[R]
Hide caps lock indicator
.TP
\f[CR]\-sm, \-\-show\-mode\f[R]
Show mode indicator
.TP
\f[CR]\-sit, \-\-show\-item\f[R]
Show items
.TP
\f[CR]\-smc, \-\-show\-match\-count\f[R]
Show match count
.TP
\f[CR]\-sla, \-\-show\-left\-arrow\f[R]
Show left arrow
.TP
\f[CR]\-sra, \-\-show\-right\-arrow\f[R]
Show right arrow
.TP
\f[CR]\-spr, \-\-show\-prompt\f[R]
Show prompt
.TP
\f[CR]\-sin, \-\-show\-input\f[R]
Show input
.TP
\f[CR]\-spt, \-\-show\-pretext\f[R]
Show pretext
.TP
\f[CR]\-spl, \-\-show\-powerline\f[R]
Show powerline
.TP
\f[CR]\-sc, \-\-show\-caret, \-\-show\-cursor\f[R]
Show caret
.TP
\f[CR]\-shl, \-\-show\-highlighting\f[R]
Show highlight
.TP
\f[CR]\-si, \-\-show\-image\f[R]
Show image
.TP
\f[CR]\-scl, \-\-show\-caps\f[R]
Show caps lock indicator
.TP
\f[CR]\-xrdb, \-\-xrdb\f[R]
Load .Xresources on runtime (X11 only)
.TP
\f[CR]\-nxrdb, \-\-no\-xrdb\f[R]
Don\[cq]t load .Xresources on runtime (X11 only)
.TP
\f[CR]\-m, \-\-monitor monitor\f[R]
Specify a monitor to run Majorna on (X11 only)
.TP
\f[CR]\-w, \-\-embed window id\f[R]
Embed Majorna inside window id (X11 only)
.TP
\f[CR]\-H, \-\-hist\-file hist file\f[R]
Specify a file to save the history to
.TP
\f[CR]\-lf, \-\-list\-file list file\f[R]
Specify a file to load entries from
.TP
\f[CR]\-ig, \-\-image\-gaps gaps\f[R]
Set image gaps to gaps
.TP
\f[CR]\-txp, \-\-text\-padding padding\f[R]
Set text padding to padding
.TP
\f[CR]\-vem, \-\-vertical\-margin margin\f[R]
Set the vertical margin to margin
.TP
\f[CR]\-hom, \-\-horizontal\-margin margin\f[R]
Set the horizontal margin to margin
.TP
\f[CR]\-lp, \-\-vertical\-padding padding\f[R]
Set the vertical padding to padding
.TP
\f[CR]\-hp, \-\-horizontal\-padding padding\f[R]
Set the horizontal padding to padding
.TP
\f[CR]\-la, \-\-left\-arrow\-symbol symbol\f[R]
Set the left arrow to symbol
.TP
\f[CR]\-ra, \-\-right\-arrow\-symbol symbol\f[R]
Set the right arrow to symbol
.TP
\f[CR]\-is, \-\-image\-size size\f[R]
Image size
.TP
\f[CR]\-it, \-\-image\-top\f[R]
Position the image at the top
.TP
\f[CR]\-ib, \-\-image\-bottom\f[R]
Position the image at the bottom
.TP
\f[CR]\-ic, \-\-image\-center\f[R]
Position the image in the center
.TP
\f[CR]\-itc, \-\-image\-topcenter\f[R]
Position the image in the top center
.TP
\f[CR]\-ir, \-\-image\-resize\f[R]
Allow Majorna to resize itself to fit the image
.TP
\f[CR]\-nir, \-\-no\-image\-resize\f[R]
Don\[cq]t allow Majorna to resize itself to fit the image
.TP
\f[CR]\-di, \-\-display\-icons\f[R]
Display the images as icons
.TP
\f[CR]\-df, \-\-display\-image\f[R]
Display the images as images in the image pane
.TP
\f[CR]\-wm, \-\-managed, \-\-x11\-client\f[R]
Spawn Majorna as a window manager controlled client/window (X11 only)
.TP
\f[CR]\-nwm, \-\-unmanaged\f[R]
Don\[cq]t spawn Majorna as a window manager controlled client/window
(X11 only)
.TP
\f[CR]\-gk, \-\-grab\-keyboard\f[R]
Grab keyboard on runtime
.TP
\f[CR]\-ngk, \-\-no\-grab\-keyboard\f[R]
Grab keyboard on runtime
.TP
\f[CR]\-cf, \-\-config\-file file\f[R]
Set config file to load to file
.TP
\f[CR]\-lcfg, \-\-load\-config\f[R]
Load majorna.configuration (\[ti]/.config/majorna/majorna.conf) on
runtime
.TP
\f[CR]\-ncfg, \-\-no\-load\-config\f[R]
Don\[cq]t load majorna.configuration
(\[ti]/.config/majorna/majorna.conf) on runtime
.TP
\f[CR]\-bf, \-\-bind\-file file\f[R]
Set bind file to load to file
.TP
\f[CR]\-lbi, \-\-load\-binds\f[R]
Exclusively load binds from file (\[ti]/.config/majorna/binds.conf) on
runtime
.TP
\f[CR]\-nlbi, \-\-no\-load\-binds\f[R]
Don\[cq]t exclusively load binds from file
(\[ti]/.config/majorna/binds.conf) on runtime
.TP
\f[CR]\-tm, \-\-theme theme\f[R]
Load theme `theme' on runtime
.TP
\f[CR]\-ltm, \-\-load\-theme\f[R]
Load theme (\[ti]/.config/majorna/theme.conf) on runtime
.TP
\f[CR]\-nltm, \-\-no\-load\-theme\f[R]
Don\[cq]t load theme (\[ti]/.config/majorna/theme.conf) on runtime
.TP
\f[CR]\-x11, \-\-x11\f[R]
Run Majorna in X11 mode
.TP
\f[CR]\-wl, \-\-wayland\f[R]
Run Majorna in Wayland mode
.TP
\f[CR]\-v, \-\-version\f[R]
Print Majorna version to stdout
.TP
\f[CR]\-rv, \-\-raw\-version\f[R]
Print raw Majorna version number to stdout
.TP
\f[CR]\-fl, \-\-feature\-list\f[R]
List the state of all features that can be toggled
.TP
\f[CR]\-fn, \-\-font font\f[R]
Set the Majorna font to font
.TP
\f[CR]\-nif, \-\-normal\-item\-foreground color\f[R]
Set the normal item foreground color
.TP
\f[CR]\-nib, \-\-normal\-item\-background color\f[R]
Set the normal item background color
.TP
\f[CR]\-nnif, \-\-normal\-next\-item\-foreground color\f[R]
Set the normal next item foreground color
.TP
\f[CR]\-nnib, \-\-normal\-next\-item\-background color\f[R]
Set the normal next item background color
.TP
\f[CR]\-sif, \-\-selected\-item\-foreground color\f[R]
Set the selected item foreground color
.TP
\f[CR]\-sib, \-\-selected\-item\-background color\f[R]
Set the selected item background color
.TP
\f[CR]\-npf, \-\-normal\-item\-priority\-foreground color\f[R]
Set the normal item (high priority) foreground color
.TP
\f[CR]\-npb, \-\-normal\-item\-priority\-background color\f[R]
Set the normal item (high priority) background color
.TP
\f[CR]\-spf, \-\-selected\-item\-priority\-foreground color\f[R]
Set the selected item (high priority) foreground color
.TP
\f[CR]\-spb, \-\-selected\-item\-priority\-background color\f[R]
Set the selected item (high priority) background color
.TP
\f[CR]\-pfg, \-\-prompt\-foreground color\f[R]
Set the prompt foreground color
.TP
\f[CR]\-pbg, \-\-prompt\-background color\f[R]
Set the prompt background color
.TP
\f[CR]\-ifg, \-\-input\-foreground color\f[R]
Set input foreground color
.TP
\f[CR]\-ibg, \-\-input\-background color\f[R]
Set input background color
.TP
\f[CR]\-ptfg, \-\-pretext\-foreground color\f[R]
Set pretext foreground color
.TP
\f[CR]\-ptbg, \-\-pretext\-background color\f[R]
Set pretext background color
.TP
\f[CR]\-mnbg, \-\-menu\-background color\f[R]
Set the menu background color
.TP
\f[CR]\-nhf, \-\-normal\-highlight\-foreground color\f[R]
Set the normal highlight foreground color
.TP
\f[CR]\-nhb, \-\-normal\-highlight\-background color\f[R]
Set the normal highlight background color
.TP
\f[CR]\-shf, \-\-selected\-highlight\-foreground color\f[R]
Set the selected highlight foreground color

docs/majorna.py
# majorna.py
# ---------
# This is the default configuration file for Majorna.
# It should be placed in ~/.config/majorna/ as majorna.py.
# If it isn't being loaded on startup, ensure Python is installed as it is not a hard dependency for Majorna.
from libmajorna import settings
from libmajorna.settings import *
from libmajorna import core

# This is a Settings object with all the default options specified.
# You can get an identical object by calling settings.get_default_config()
# It is safe to remove any values you don't want to explicitly specify.
settings = Settings(
    window=Window(
        position=WindowPosition.Center, # Position to place Majorna in
        monitor=-1, # Monitor to place Majorna on. -1 = auto
        protocol=PreferredProtocol.Wayland, # Preferred protocol
        border_size=0, # Size of the border around the window
        margin_vertical=0, # Vertical margin
        margin_horizontal=0, # Horizontal margin
        padding_vertical=0, # Vertical padding
        padding_horizontal=0, # Horizontal padding
        dimensions=Dimensions( # Window positioning/dimensions, 0 = auto
            x=0,
            y=0,
            width=0
        ),
        center_width=1000, # Window width when position=WindowPosition.Center
        grab_input=True, # Grab keyboard and mouse input
        enable_transparency=True # Enable transparency (Requires a compositor on X11)
    ),
    x11=X11(
        class_name='majorna', # Class name
        dock_type=True, # Set the _NET_WM_WINDOW_TYPE_DOCK property
        managed=False, # Create the window as a standard window
        xresources=True # Enable xrdb/.Xresources support for loading properties
    ),
    text=Text(
        font='Noto Sans Mono 10', # Preferred font (syntax: <font name> <font size>). Bitmap fonts are not supported.
        left_arrow='<', # Symbol to represent the arrow on the left side, when items are present on the left.
        right_arrow='>', # Symbol to represent the arrow on the right side, when items are present on the right.
        password='*', # Symbol to use in place of each character when password mode is enabled.
        prompt='', # Default prompt text
        pretext='', # Default pretext
        input='', # Default input
        normal_mode='Normal', # Normal mode text
        insert_mode='Insert', # Insert mode text
        regex_mode='Regex', # Regex mode text
        caps_lock_on='Caps Lock', # Caps Lock on text
        caps_lock_off='', # Caps Lock off text
        padding=0, # Padding around all drawn text
        normal_item_padding=0, # Padding around normal items
        selected_item_padding=0, # Padding around selected items
        priority_item_padding=0 # Padding around high priority items
    ),
    appearance=Appearance(
        color=AppearanceColor(
            item=ColorItem(
                normal_foreground='#bbbbbb',# Normal item foreground color
                normal_background='#110f1f', # Normal item background color
                normal_foreground_second='#bbbbbb', # Normal item foreground color (every second item)
                normal_background_second='#110f1f', # Normal item background color (every second item)
                selected_foreground='#110f1f', # Selected item foreground color
                selected_background='#8e93c2', # Selected item background color
                selected_foreground_second='#110f1f', # Selected item foreground color (every second item)
                selected_background_second='#8e93c2', # Selected item background color (every second item)
                marked_foreground='#110f1f', # Marked item foreground color
                marked_background='#8e93c2', # Marked item background color
                normal_foreground_priority='#bbbbbb', # Normal, high priority foreground color
                normal_background_priority='#110f1f', # Normal, high priority background color
                selected_foreground_priority='#110f1f', # Selected, high priority, foreground color
                selected_background_priority='#8e93c2', # Selected, high priority, background color
            ),
            input=InputColor(
                foreground='#eeeeee', # Input foreground color
                background='#110f1f', # Input background color
            ),
            pretext=PretextColor(
                foreground='#999888', # Pretext foreground color
                background='#110f1f', # Pretext background color
            ),
            prompt=PromptColor(
                foreground='#110f1f', # Prompt foreground color
                background='#c66e5b', # Prompt background color
            ),
            left_arrow=LeftArrowColor(
                foreground='#bbbbbb', # Left arrow foreground color
                background='#110f1f', # Left arrow background color
            ),
            right_arrow=RightArrowColor(
                foreground='#bbbbbb', # Right arrow foreground color
                background='#110f1f', # Right arrow background color
            ),
            highlighting=HighlightColor(
                normal_foreground='#ffffff', # Normal highlighting foreground color
                normal_background='#000000', # Normal highlighting background color
                selected_foreground='#ffffff', # Selected highlighting foreground color
                selected_background='#000000', # Selected highlighting background color
            ),
            match_counter=MatchCounterColor(
                foreground='#110f1f', # Match counter foreground color
                background='#eba62a', # Match counter background color
            ),
            mode_indicator=ModeIndicatorColor(
                foreground='#110f1f', # Mode indicator foreground color
                background='#92c94b', # Mode indicator background color
            ),
            caps_lock_indicator=CapsLockIndicatorColor(
                foreground='#110f1f', # Caps Lock indicator foreground color
                background='#f8d3e6', # Caps Lock indicator background color
            ),
            caret=CaretColor(
                foreground='#ffffff', # Caret foreground color
                background='#110f1f', # Caret background color
            ),
            general=GeneralColor(
                menu='#110f1f', # Menu color
                border='#8e93c2', # Border color
            ),
            sgr=[
                '#20201d', # SGR color 0
                '#d73737', # SGR color 1
                '#60ac39', # SGR color 2
                '#cfb017', # SGR color 3
                '#6684e1', # SGR color 4
                '#b854d4', # SGR color 5
                '#1fad83', # SGR color 6
                '#a6a28c', # SGR color 7
                '#7d7a68', # SGR color 8
                '#d73737', # SGR color 9
                '#60ac39', # SGR color 10
                '#cfb017', # SGR color 11
                '#6684e1', # SGR color 12
                '#b854d4', # SGR color 13
                '#1fad83', # SGR color 14
                '#fefbec', # SGR color 15
            ]
        ),
        transparency=Transparency(
            item=ItemTransparency(
                normal_foreground=255, # Normal item foreground alpha
                normal_background=222, # Normal item background alpha
                normal_foreground_second=255, # Normal item foreground alpha (every second item)
                normal_background_second=222, # Normal item background alpha (every second item)
                selected_foreground=255, # Selected item foreground alpha
                selected_background=222, # Selected item background alpha
                selected_foreground_second=255, # Selected item foreground alpha (every second item)
                selected_background_second=222, # Selected item background alpha (every second item)
                marked_foreground=255, # Marked item foreground alpha
                marked_background=222, # Marked item background alpha
                normal_foreground_priority=255, # Normal, high priority foreground alpha
                normal_background_priority=222, # Normal, high priority background alpha
                selected_foreground_priority=255, # Selected, high priority foreground alpha
                selected_background_priority=222, # Selected, high priority background alpha
            ),
            input=InputTransparency(
                foreground=255, # Input foreground alpha
                background=222, # Input background alpha
            ),
            pretext=PretextTransparency(
                foreground=255, # Pretext foreground alpha
                background=222, # Pretext background alpha
            ),
            prompt=PromptTransparency(
                foreground=255, # Prompt foreground alpha
                background=222, # Prompt background alpha
            ),
            left_arrow=LeftArrowTransparency(
                foreground=255, # Left arrow foreground alpha
                background=222, # Left arrow background alpha
            ),
            right_arrow=RightArrowTransparency(
                foreground=255, # Right arrow foreground alpha
                background=222, # Right arrow background alpha
            ),
            highlighting=HighlightTransparency(
                normal_foreground=255, # Normal highlighting foreground alpha
                normal_background=222, # Normal highlighting background alpha
                selected_foreground=255, # Selected highlighting foreground alpha
                selected_background=222, # Selected highlighting background alpha
            ),
            match_counter=MatchCounterTransparency(
                foreground=255, # Match counter foreground alpha
                background=222, # Match counter background alpha
            ),
            mode_indicator=ModeIndicatorTransparency(
                foreground=255, # Mode indicator foreground alpha
                background=222, # Mode indicator background alpha
            ),
            caps_lock_indicator=CapsLockIndicatorTransparency(
                foreground=255, # Caps Lock indicator foreground alpha
                background=222, # Caps Lock indicator background alpha
            ),
            caret=CaretTransparency(
                foreground=255, # Caret foreground alpha
                background=222, # Caret background alpha
            ),
            general=GeneralTransparency(
                menu=222, # Menu alpha
                border=222, # Border alpha
            )
        ),
        hide=Hide(
            input=False, # Hide input
            pretext=False, # Hide pretext
            prompt=False, # Hide prompt
            left_arrow=False, # Hide left arrow
            right_arrow=False, # Hide right arrow
            items=False, # Hide items
            powerline=False, # Hide powerline
            caret=False, # Hide caret
            highlighting=False, # Hide highlighting
            match_counter=False, # Hide match counter
            mode_indicator=False, # Hide mode indicator
            caps_lock_indicator=False, # Hide caps lock indicator
            images=False, # Hide items
        ),
        powerline=Powerline(
            style=PowerlineStyling(
                prompt_style=PowerlineStyle.HalfCircle, # Prompt powerline style
                match_counter_style=PowerlineStyle.HalfCircle, # Match counter powerline style
                mode_indicator_style=PowerlineStyle.HalfCircle, # Mode indicator powerline style
                caps_lock_indicator_style=PowerlineStyle.HalfCircle, # Caps lock indicator powerline style
                item_style=PowerlineStyle.HalfCircle, # Item powerline style
            ),
            enable=PowerlineEnable(
                prompt=True, # Enable powerline for the prompt
                match_counter=True, # Enable powerline for the match counter
                mode_indicator=True, # Enable powerline for the mode indicator
                caps_lock_indicator=True, # Enable powerline for the caps lock indicator
                item=True, # Enable powerline for the items
            )
        ),
        line=Line(
            item_position=ItemPosition.Bottom, # Item positioning
            height=1, # Item height, 0 = auto
            lines=0, # Default number of lines
            columns=10, # Default number of columns
            minimum_lines=0, # Minimum number of lines allowed
            indent_items_to_prompt=False, # Indent items to prompt width when lines > 0
        ),
        input=Input(
            width=0.3, # Input area width
        ),
        caret=Caret(
            dimensions=CaretDimensions(
                width=0, # Caret width
                height=0, # Caret height
                padding=0, # Caret padding
            )
        ),
        color_items=True, # Color items
        enable_sgr_sequences=True, # Enable SGR sequences
    ),
    behavior=Behavior(
        display=Display(
            sort=True, # Sort items
            preselected=0, # Preselect an item index
        ),
        matching=Matching(
            case_sensitive=False, # Case sensitive matching
            fuzzy_matching=True, # Fuzzy finding
            regex_matching=False, # Regex matching
        ),
        permissions=Permissions(
            mark=True, # Allow marking items
            typing=True, # Allow typing
            password=False, # Hide the input and any items
        ),
        output=Output(
            print_index=False, # Print out the index instead of the actual item text
            incremental=False, # Print out on every keypress
        ),
        miscellaneous=Miscellaneous(
            fast=True, # Grab keyboard before matching
            delimiters=' /?\\"&[]', # Delimiters for word deletion
            list_file='', # File to read items from
            start_mode=Mode.Normal, # Mode to start Majorna in
            force_insert_mode=True, # Force insert mode (i.e. disable normal mode)
            scroll_distance=512, # Scroll distance required for a scroll action to count (Wayland only)
        ),
        pango=Pango(
            item=True, # Enable Pango syntax for the items
            prompt=True, # Enable Pango syntax for the prompt
            input=True, # Enable Pango syntax for the input
            pretext=True, # Enable Pango syntax for the pretext
            left_arrow=False, # Enable Pango syntax for the left arrow
            right_arrow=False, # Enable Pango syntax for the right arrow
            match_counter=False, # Enable Pango syntax for the match counter
            mode_indicator=False, # Enable Pango syntax for the mode indicator
            caps_lock_indicator=False, # Enable Pango syntax for the caps lock indicator
            password=False, # Enable Pango syntax for the password input
        )
    ),
    history=History(
        max_number_of_entries=64, # Max number of entries to save in the history file
        save_duplicates=False, # Save duplicates to the history file
    ),
    image=Image(
        dimensions=ImageDimensions(
            width=200, # Default image width
            height=200, # Default image height
            gaps=0, # Default image gaps
        ),
        position=ImagePosition.Top, # Default image position
        type=ImageType.Image, # Default image type
        cache=True, # Cache images
        max_size_to_cache=512, # Max image size to cache
        resize_to_fit=True, # Resize the image to fit
    ),
    filesystem=Filesystem(
        paths=FilesystemPaths(
            fifo='/tmp/majorna.fifo', # FIFO path (Deprecated)
            screenshot='%h/Screenshots/majorna-%d-%t.png', # Screenshot path (%h = Home directory, %d = Date, %t = Time)
            image_cache_directory='default', # Image cache directory
        )
    ),
    keys=[
        Key(
            mode=KeyMode.Any,
            modifier='None',
            key='Enter',
            function='select_item',
            argument=1),
        Key(
            mode=KeyMode.Any,
            modifier='Shift',
            key='Enter',
            function='select_item',
            argument=0),
        Key(
            mode=KeyMode.Any,
            modifier='Ctrl',
            key='Enter',
            function='mark_item',
            argument=0),
        Key(
            mode=KeyMode.Any,
            modifier='None',
            key='Tab',
            function='complete',
            argument=0),
        Key(
            mode=KeyMode.Any,
            modifier='Ctrl',
            key='v',
            function='paste',
            argument=2),
        Key(
            mode=KeyMode.Any,
            modifier='None',
            key='Backspace',
            function='backspace',
            argument=0),
        Key(
            mode=KeyMode.Any,
            modifier='Ctrl',
            key='Backspace',
            function='delete_word',
            argument=0),
        Key(
            mode=KeyMode.Any,
            modifier='Ctrl+Shift',
            key='p',
            function='set_profile',
            argument=0),
        Key(
            mode=KeyMode.Any,
            modifier='Ctrl',
            key='=',
            function='set_image_size',
            argument=10),
        Key(
            mode=KeyMode.Any,
            modifier='Ctrl',
            key='-',
            function='set_image_size',
            argument=-10),
        Key(
            mode=KeyMode.Any,
            modifier='None',
            key='Up',
            function='move_up',
            argument=0),
        Key(
            mode=KeyMode.Any,
            modifier='None',
            key='Down',
            function='move_down',
            argument=0),
        Key(
            mode=KeyMode.Any,
            modifier='None',
            key='Left',
            function='move_left',
            argument=0),
        Key(
            mode=KeyMode.Any,
            modifier='None',
            key='Right',
            function='move_right',
            argument=0),
        Key(
            mode=KeyMode.Any,
            modifier='Ctrl',
            key='u',
            function='move_up',
            argument=5),
        Key(
            mode=KeyMode.Any,
            modifier='Ctrl',
            key='d',
            function='move_down',
            argument=5),
        Key(
            mode=KeyMode.Any,
            modifier='Ctrl',
            key='h',
            function='view_history',
            argument=0),
        Key(
            mode=KeyMode.Any,
            modifier='None',
            key='Esc',
            function='quit',
            argument=0),
        Key(
            mode=KeyMode.Any,
            modifier='Ctrl',
            key='p',
            function='navigate_history',
            argument=-1),
        Key(
            mode=KeyMode.Any,
            modifier='Ctrl',
            key='n',
            function='navigate_history',
            argument=1),
        Key(
            mode=KeyMode.Any,
            modifier='None',
            key='m',
            function='screenshot',
            argument=0)
    ],

    mouse=[
        Mouse(
            click=ClickType.Input,
            button=ButtonType.LeftClick,
            function='clear',
        ),
        Mouse(
            click=ClickType.Prompt,
            button=ButtonType.LeftClick,
            function='clear',
        ),
        Mouse(
            click=ClickType.ModeIndicator,
            button=ButtonType.LeftClick,
            function='toggle_mode'
        ),
        Mouse(
            click=ClickType.MatchCounter,
            button=ButtonType.LeftClick,
            function='view_history',
        ),
        Mouse(
            click=ClickType.Item,
            button=ButtonType.LeftClick,
            function='select_hover',
        ),
        Mouse(
            click=ClickType.Item,
            button=ButtonType.RightClick,
            function='mark_hover',
        ),
        Mouse(
            click=ClickType.Any,
            button=ButtonType.ScrollUp,
            function='move_prev',
        ),
        Mouse(
            click=ClickType.Any,
            button=ButtonType.ScrollDown,
            function='move_next'
        ),
    ]
)

# Import the settings dataclass we created
core.load_settings(settings)
